enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/RaftInstance.java

```diff
@@ -187,21 +185,75 @@ public void setTargetMembershipSet( Set<MEMBER> targetMembers )
     }
 
     @Override
-    public MEMBER getLeader() throws NoLeaderTimeoutException
+    public MEMBER getLeader() throws NoLeaderFoundException
     {
-        long leaderWaitEndTime = leaderWaitTimeout + clock.currentTimeMillis();
-        while ( state.leader() == null && (clock.currentTimeMillis() < leaderWaitEndTime) )
+        return getLeader( leaderWaitTimeout );
+    }
+
+    @Override
+    public MEMBER getLeader( long timeoutMillis ) throws NoLeaderFoundException
+    {
+        try
         {
-            LockSupport.parkNanos( MILLISECONDS.toNanos( electionTimeout ) / 2 );
+            return getLeader0( timeoutMillis );
         }
-
-        if ( state.leader() == null )
+        catch( NoLeaderFoundException e )
         {
             leaderNotFoundMonitor.increment();
-            throw new NoLeaderTimeoutException();
+            throw e;
+        }
+    }
+
+    public MEMBER getLeader0( long timeoutMillis ) throws NoLeaderFoundException
+    {
+        MEMBER leaderAlias = state.leader();
+        if ( leaderAlias == null )
+        {
+            if ( timeoutMillis > 0 )
+            {
+                long endTimeMillis = timeoutMillis + clock.currentTimeMillis();
+                try
+                {
+                    leaderAlias = waitForLeader( endTimeMillis );
+                }
+                catch ( InterruptedException e )
+                {
+                    Thread.currentThread().interrupt();
+                    throw new NoLeaderFoundException( e );
+                }
+                catch ( TimeoutException e )
+                {
+                    throw new NoLeaderFoundException( e );
+                }
+
+                return leaderAlias;
+            }
+            else
+            {
+                throw new NoLeaderFoundException();
+            }
+        }
+
+        return leaderAlias;
+    }
+
+    private synchronized MEMBER waitForLeader( long endTimeMillis ) throws InterruptedException, TimeoutException
```

Comment:
That sort of polling was one of the things I am trying to get rid off. And the sleep of timeout/100 just looks arbitrary.